src/coreclr/jit/codegenarmarch.cpp

```diff
@@ -2558,12 +2576,20 @@ void CodeGen::genCallInstruction(GenTreeCall* call)
                ((call->IsVirtualStubRelativeIndir()) && (call->gtEntryPoint.accessType == IAT_VALUE)));
 #endif // FEATURE_READYTORUN
         assert(call->gtControlExpr == nullptr);
-        assert(!call->IsTailCall());
 
         regNumber tmpReg = call->GetSingleTempReg();
-        regNumber callAddrReg =
-            call->IsVirtualStubRelativeIndir() ? compiler->virtualStubParamInfo->GetReg() : REG_R2R_INDIRECT_PARAM;
-        GetEmitter()->emitIns_R_R(ins_Load(TYP_I_IMPL), emitActualTypeSize(TYP_I_IMPL), tmpReg, callAddrReg);
+        // For fast tailcalls we have already loaded the call target when processing the call node.
+        if (!call->IsFastTailCall())
+        {
+            regNumber callAddrReg =
+                call->IsVirtualStubRelativeIndir() ? compiler->virtualStubParamInfo->GetReg() : REG_R2R_INDIRECT_PARAM;
+            GetEmitter()->emitIns_R_R(ins_Load(TYP_I_IMPL), emitActualTypeSize(TYP_I_IMPL), tmpReg, callAddrReg);
+        }
+        else
+        {
+            // Register where we save call address in should not be overridden by epilog.
+            assert((tmpReg & (RBM_INT_CALLEE_TRASH & ~RBM_LR)) == tmpReg);
```

Comment:
I'm confused about the role of `tmpReg` in the fast tail call case. How does it end up having the right contents?

Reply:
It happens in `genCall` here: https://github.com/jakobbotsch/runtime/blob/9f70cc42d73467c15a458ee9774589271d721536/src/coreclr/jit/codegenarmarch.cpp#L2340-L2375

For normal calls we call `genCall` which then calls `genCallInstruction` that takes care to generate the code to load the call target and do the call.
For tailcalls we instead generate the code to load the call target in `genCall`, but this is the last thing that happens when we see the `GenTreeCall`  node. The remaining work happens during epilog generation which also calls `genCallInstruction`.

Reply:
So you could pass REG_NA for the tail call case?

I guess it's `GetSingleTempReg` that is confusing me. It seems odd to "allocate" a temp reg without altering state.

Reply:
> It seems odd to "allocate" a temp reg without altering state.

Not sure I understand. The temp reg is allocated during RA for this specific optimization we do on ARM/ARM64, where we have no target node and need an extra register to store the call target loaded from the indirection cell. We still need it for the tail call case.

Reply:
Right, not that it's not ultimately needed, but that the value of `tmpReg` passed here to GenCall is not used, instead we go call `GetSingleTempReg` again.

It's not that important, I just found it confusing to follow how the value gets from one place to another.

Reply:
This code here in `genCallInstruction` is called last. `genCall` also calls `GetSingleTempReg`, but it adds the register back to `gtRsvdRegs` so that this call will succeed:
https://github.com/jakobbotsch/runtime/blob/9f70cc42d73467c15a458ee9774589271d721536/src/coreclr/jit/codegenarmarch.cpp#L2369-L2370

Reply:
Thanks, makes more sense now.